# Converting Values

Converting between different numerical and data representations is an essential skill in programming, allowing for efficient data processing, interoperability, and the ability to build versatile applications. This guide will explore how to validate, convert, encode, and decode various formats, making your programs both robust and flexible. These concepts also serve as a foundation for implementing more complex systems in SplashKit, such as cryptographic utilities or network tools.

## Validation Functions

Validation is the first step in any reliable conversion process. Ensuring input values conform to their respective formats reduces errors and creates a solid foundation for processing data. For example, binary numbers must consist only of `0`s and `1`s, while hexadecimal values use a combination of numeric and alphabetic characters. In this section, we’ll explore how to check the correctness of binary, decimal, hexadecimal, and octal values, ensuring all inputs are clean and error-free before proceeding with conversions.

:::note
The following validation functions are used in the conversion operations described later in this guide. They check whether a given string is a valid representation of a binary, decimal, hexadecimal, or octal number, respectively.
:::

### Is Binary

This function checks if a given string is a valid binary number, ensuring it contains only the characters `0` and `1`. It also verifies that the input is not empty, returning false for any invalid or empty input.

In [71]:
def is_binary(binary_string):
    return all(character in '01' for character in binary_string) and bool(binary_string)

We can use this function within many other functions, or even to just check if a string is a binary number:

In [72]:
binary_string1 = "1003300101"
print(F"Value {binary_string1} is a {is_binary(binary_string1)} binary value")

binary_string2 = "10010101"
print(F"Value {binary_string2} is a {is_binary(binary_string2)} binary value")

Value 1003300101 is a False binary value
Value 10010101 is a True binary value


### Is Hexadecimal

This function validates whether a string represents a valid hexadecimal number. It checks that each character falls within the ranges `0-9`, `A-F`, or `a-f`, while also ensuring the input is not empty.

In [73]:
def is_hex(hex_string):
    return all(character in '0123456789ABCDEFabcdef' for character in hex_string) and bool(hex_string)

We can use this function to verify hexadecimal strings before converting them to other formats:

In [74]:
hex_string1 = "1A2B3C4D5F"
print(f"Value {hex_string1} is a {is_hex(hex_string1)} hex value") # True

hex_string2 = "1A2B3C4D5G"
print(f"Value {hex_string2} is a {is_hex(hex_string2)} hex value") # False

Value 1A2B3C4D5F is a True hex value
Value 1A2B3C4D5G is a False hex value


### Is Octal

This function determines if a string is a valid octal number by ensuring all characters are within the range `0-7`. It also returns false if the input string is empty.

In [75]:
def is_octal(octal_string):
    return all(character in '01234567' for character in octal_string) and bool(octal_string)

We can use this function to validate octal strings before converting them to other formats:

In [76]:
octal_string1 = "1234567"
print(f"Value {octal_string1} is a {is_octal(octal_string1)} octal value") # True

octal_string2 = "12345678"
print(f"Value {octal_string2} is a {is_octal(octal_string2)} octal value") # False

Value 1234567 is a True octal value
Value 12345678 is a False octal value


## Basic Conversion Operations

Conversions between number systems are fundamental to understanding how data is represented in programming. Whether you’re transforming binary to decimal, decimal to hexadecimal, or even converting between less common systems like octal and base-64, these operations provide insight into the underlying mechanics of data storage and manipulation. This section will introduce methods for performing these conversions step by step, helping you understand the relationships between different numerical systems and how to implement them efficiently in code.

### Decimal to Binary

Converting a decimal number to binary involves dividing the number by 2 repeatedly and recording the remainder for each division. These remainders, read in reverse order, form the binary representation of the decimal number. For example, to convert the decimal number 10 to binary, you would perform the following steps: $10 ÷ 2 = 5$ remainder $0$, $5 ÷ 2 = 2$ remainder $1$, $2 ÷ 2 = 1$ remainder $0$, and $1 ÷ 2 = 0$ remainder $1$. Reading the remainders from bottom to top, you get the binary number $1010$. This process can be implemented programmatically to convert any decimal number to its binary equivalent.

:::tip
This could be useful for encoding data into binary formats like Base64, performing low-level programming tasks, or representing numbers in digital circuits.
:::

In [77]:
def dec_to_bin(decimal_value):
    if decimal_value == 0:
        return "0"
    
    binary_string = ""
    while decimal_value > 0:
        binary_string = ("1" if decimal_value & 1 else "0") + binary_string
        decimal_value >>= 1
    return binary_string

In [78]:
decimal_value = 10
binary_string = dec_to_bin(decimal_value)
print("Binary value:", binary_string)

Binary value: 1010


If we change the input value to $22$, the output will be:

In [79]:
decimal_value = 22
binary_string = dec_to_bin(decimal_value)
print("Binary value:", binary_string)

Binary value: 10110


### Binary to Decimal

Converting a binary number to decimal involves multiplying each digit by $2^n$, where $n$ is the position of the digit from the rightmost bit, starting at $0$. For example, to convert the binary number $1010$ to decimal, you would calculate $1 \times 2^3 + 0 \times 2^2 + 1 \times 2^1 + 0 \times 2^0 = 10$. This process can be implemented programmatically to convert any binary number to its decimal equivalent.

In [80]:
def bin_to_dec(binary_string):
    if not is_binary(binary_string):
        return 0

    decimal_value = 0
    for i, character in enumerate(reversed(binary_string)):
        if character == '1':
            decimal_value += (1 << i)
    return decimal_value

In [81]:
binary_string = "1010"
decimal_value = bin_to_dec(binary_string)
print("Decimal value:", decimal_value)

Decimal value: 10


If we change the input value to $10110$, the output will be:

In [82]:
binary_string = "10110"
decimal_value = bin_to_dec(binary_string)
print("Decimal value:", decimal_value)

Decimal value: 22


### Hexadecimal to Binary

Converting a hexadecimal number to binary involves converting each hexadecimal digit to its binary equivalent. For example, the hexadecimal number $A3$ can be converted to binary by converting $A$ to $1010$ and $3$ to $0011$, resulting in the binary number $10100011$. This process can be implemented programmatically to convert any hexadecimal number to its binary equivalent.

In [83]:
def hex_to_bin(hex_string):
    if not is_hex(hex_string):
        return ""
    
    binary_string = ""
    for hex_character in hex_string:
        hex_value = int(hex_character, 16)
        for i in range(3, -1, -1):
            binary_string += "1" if (hex_value >> i) & 1 else "0"
    
    if len(hex_string) == 1:
        first_one = binary_string.find('1')
        return "0" if first_one == -1 else binary_string[first_one:]
    
    return binary_string

In [84]:
hex_string = "A3"
binary_string = hex_to_bin(hex_string)
print("Binary value:", binary_string)

Binary value: 10100011


If we change the input value to $FA$, the output will be:

In [85]:
hex_string = "FA"
binary_string = hex_to_bin(hex_string)
print("Binary value:", binary_string)

Binary value: 11111010


### Binary to Hexadecimal

Converting a binary number to hexadecimal involves grouping the binary digits into sets of four, starting from the rightmost bit. Each group of four bits corresponds to a single hexadecimal digit, allowing you to convert the binary number to its hexadecimal equivalent. For example, the binary number $10100011$ can be grouped as $1010$ and $0011$, which correspond to the hexadecimal digits $A$ and $3$, respectively. This process can be implemented programmatically to convert any binary number to its hexadecimal equivalent.

In [86]:
def bin_to_hex(binary_string):
    if not is_binary(binary_string):
        return ""
    
    length = len(binary_string)
    padding = (4 - (length % 4)) % 4
    padded_binary_string = "0" * padding + binary_string
    
    hex_string = ""
    for i in range(0, len(padded_binary_string), 4):
        hex_value = int(padded_binary_string[i:i+4], 2)
        hex_string += hex(hex_value)[2:].upper()
    return hex_string

In [87]:
binary_string = "10100011"
hex_string = bin_to_hex(binary_string)
print("Hexadecimal value:", hex_string)

Hexadecimal value: A3


If we change the input value to $1100101011$, the output will be:

In [88]:
binary_string = "1100101011"
hex_string = bin_to_hex(binary_string)
print("Hexadecimal value:", hex_string)

Hexadecimal value: 32B


### Decimal to Octal

Converting a decimal number to octal involves dividing the number by 8 repeatedly and recording the remainder for each division. These remainders, read in reverse order, form the octal representation of the decimal number. For example, to convert the decimal number 10 to octal, you would perform the following steps: $10 ÷ 8 = 1$ remainder $2$, and $1 ÷ 8 = 0$ remainder $1$. Reading the remainders from bottom to top, you get the octal number $12$. This process can be implemented programmatically to convert any decimal number to its octal equivalent.

In [89]:
def dec_to_oct(decimal_value):
    if decimal_value == 0:
        return "0"
    
    octal_string = ""
    while decimal_value > 0:
        octal_string = str(decimal_value % 8) + octal_string
        decimal_value //= 8
    return octal_string

In [90]:
decimal_value = 10
octal_string = dec_to_oct(decimal_value)
print("Octal value:", octal_string)

Octal value: 12


If we change the input value to $87$, the output will be:

In [91]:
decimal_value = 87
octal_string = dec_to_oct(decimal_value)
print("Octal value:", octal_string)

Octal value: 127


### Octal to Decimal

Converting an octal number to decimal involves multiplying each digit by $8^n$, where $n$ is the position of the digit from the rightmost bit, starting at $0$. For example, to convert the octal number 12 to decimal, you would calculate $1 \times 8^1 + 2 \times 8^0 = 10$. This process can be implemented programmatically to convert any octal number to its decimal equivalent.

In [92]:
def oct_to_dec(octal_string):
    if not is_octal(octal_string):
        return 0

    decimal_value = 0
    for i, character in enumerate(reversed(octal_string)):
        decimal_value += int(character) * (8 ** i)
    return decimal_value

In [93]:
octal_string = "12"
decimal_value = oct_to_dec(octal_string)
print("Decimal value:", decimal_value)

Decimal value: 10


If we change the input value to $26$, the output will be:

In [94]:
octal_string = "26"
decimal_value = oct_to_dec(octal_string)
print("Decimal value:", decimal_value)

Decimal value: 22


### Octal to Binary

Converting an octal number to binary involves converting each octal digit to its binary equivalent. For example, the octal number $12$ can be converted to binary by converting $1$ to $001$ and $2$ to $010$, resulting in the binary number $001010$. This process can be implemented programmatically to convert any octal number to its binary equivalent.

In [95]:
def oct_to_bin(octal_string):
    if not is_octal(octal_string):
        return ""
    
    binary_string = ""
    for octal_character in octal_string:
        octal_value = int(octal_character)
        for i in range(2, -1, -1):
            binary_string += "1" if (octal_value >> i) & 1 else "0"
    
    first_one = binary_string.find('1')
    return "0" if first_one == -1 else binary_string[first_one:]

In [96]:
octal_string = "12"
binary_string = oct_to_bin(octal_string)
print("Binary value:", binary_string)

Binary value: 1010


If we change the input value to $26$, the output will be:

In [97]:
octal_string = "26"
binary_string = oct_to_bin(octal_string)
print("Binary value:", binary_string)

Binary value: 10110


### Binary to Octal

Converting a binary number to octal involves grouping the binary digits into sets of three, starting from the rightmost bit. Each group of three bits corresponds to a single octal digit, allowing you to convert the binary number to its octal equivalent. For example, the binary number $101010$ can be grouped as $101$ and $010$, which correspond to the octal digits $5$ and $2$, respectively. This process can be implemented programmatically to convert any binary number to its octal equivalent.

In [98]:
def bin_to_oct(binary_string):
    if not is_binary(binary_string):
        return ""
    
    padding = (3 - (len(binary_string) % 3)) % 3
    padded_binary_string = "0" * padding + binary_string
    
    octal_string = ""
    for i in range(0, len(padded_binary_string), 3):
        octal_value = int(padded_binary_string[i:i+3], 2)
        octal_string += str(octal_value)
    
    first_non_zero = octal_string.find('1')
    return "0" if first_non_zero == -1 else octal_string[first_non_zero:]

In [99]:
binary_string = "101010"
octal_string = bin_to_oct(binary_string)
print("Octal value:", octal_string)

Octal value: 0


If we change the input value to $101110010$, the output will be:

In [100]:
binary_string = "101110010"
octal_string = bin_to_oct(binary_string)
print("Octal value:", octal_string)

Octal value: 0


### Hexadecimal to Octal

Converting a hexadecimal number to octal involves converting the hexadecimal number to binary and then converting the binary number to octal. This process can be implemented programmatically to convert any hexadecimal number to its octal equivalent.

:::note
This function relies on the `hex_to_bin` and the `bin_to_oct` functions defined above.
:::

In [101]:
def hex_to_oct(hex_string):
    if not is_hex(hex_string):
        return ""
    
    binary_string = hex_to_bin(hex_string)
    return bin_to_oct(binary_string)

In [102]:
hex_string = "A3"
octal_string = hex_to_oct(hex_string)
print("Octal value:", octal_string)

Octal value: 0


If we change the input value to $FAA$, the output will be:

In [103]:
hex_string = "FAA"
octal_string = hex_to_oct(hex_string)
print("Octal value:", octal_string)

Octal value: 0


### Octal to Hexadecimal

Converting an octal number to hexadecimal involves converting the octal number to binary and then converting the binary number to hexadecimal. This process can be implemented programmatically to convert any octal number to its hexadecimal equivalent.

:::note
This function relies on the `oct_to_bin` and `bin_to_hex` functions defined above.
:::

In [104]:
def oct_to_hex(octal_string):
    if not is_octal(octal_string):
        return ""
    
    binary_string = oct_to_bin(octal_string)
    return bin_to_hex(binary_string)

In [105]:
octal_string = "12"
hex_string = oct_to_hex(octal_string)
print("Hexadecimal value:", hex_string)

Hexadecimal value: A


If we change the input value to $26$, the output will be:

In [109]:
octal_string = "26"
hex_string = oct_to_hex(octal_string)
print("Hexadecimal value:", hex_string)

Hexadecimal value: 16
